File: frontend/src/mocks/mockApi.ts
import type { Bus, Empresa, ParadaRuta, UsuarioPerfil, VentaRequest, ViajeOperador } from '@/shared/api/types';
import {
  MOCK_DELAY_MS,
  fechaHoyIso,
  mockBuses,
  mockDetalleCooperativa,
  mockDetalleViajeOperador,
  mockDetalleViajePublico,
  mockEmpresas,
  mockIngresos,
  mockManifiesto,
  mockOcupacion,
  mockOperadores,
  mockParadas,
  mockPerfil,
  mockResumenPlataforma,
  mockVentaResponse,
  mockViajesDisponibles,
  mockViajesOperador,
} from './mockData';

function delay(ms = MOCK_DELAY_MS): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function parsePath(fullPath: string): { pathname: string; searchParams: URLSearchParams } {
  const qIndex = fullPath.indexOf('?');
  const pathname = qIndex >= 0 ? fullPath.slice(0, qIndex) : fullPath;
  const query = qIndex >= 0 ? fullPath.slice(qIndex + 1) : '';
  return { pathname, searchParams: new URLSearchParams(query) };
}

function matchId(pathname: string, prefix: string): number | null {
  const m = pathname.match(new RegExp(`^${prefix.replace(/\//g, '\\/')}(\\d+)`));
  return m ? Number(m[1]) : null;
}

function filterViajes(empresaId: number | null, fecha: string, origen?: string | null): ViajeOperador[] {
  let list = mockViajesOperador(fecha);
  if (empresaId != null) list = list.filter((v) => v.empresaId === empresaId);
  if (origen) list = list.filter((v) => v.origen === origen);
  return list;
}

function filterBuses(empresaId: number | null, sede?: string | null): Bus[] {
  let list = mockBuses;
  if (empresaId != null) list = list.filter((b) => b.empresaId === empresaId);
  if (sede) list = list.filter((b) => b.sede === sede);
  return list;
}

export async function handleMockRequest<T>(fullPath: string, options: RequestInit = {}): Promise<T> {
  await delay();

  const method = (options.method ?? 'GET').toUpperCase();
  const { pathname, searchParams } = parsePath(fullPath);
  const body = options.body ? (JSON.parse(options.body as string) as Record<string, unknown>) : null;

  // --- Auth / perfil ---
  if (pathname === '/api/usuarios/me') {
    return mockPerfil as T;
  }

  if (pathname === '/api/usuarios') {
    const empresaId = searchParams.get('empresaId');
    const list = empresaId
      ? mockOperadores.filter((o) => o.empresaId === Number(empresaId))
      : mockOperadores;
    return list as T;
  }

  if (pathname.startsWith('/api/usuarios/') && method === 'PATCH') {
    const id = matchId(pathname, '/api/usuarios/');
    const user = mockOperadores.find((o) => o.id === id) ?? mockPerfil;
    return { ...user, ...body } as T;
  }

  if (pathname === '/api/usuarios' && method === 'POST') {
    const nuevo: UsuarioPerfil = {
      id: 99,
      empresaId: Number(body?.empresaId ?? 1),
      empresaNombre: 'Wendelyn Transporte',
      nombreUsuario: String(body?.nombreUsuario ?? 'nuevo.operador'),
      emailLogin: String(body?.email ?? 'nuevo@demo.com'),
      nombreCompleto: String(body?.nombreCompleto ?? 'Operador Demo'),
      sede: String(body?.sede ?? 'Bluefields'),
      activo: true,
      roles: (body?.roles as string[]) ?? ['CAJERO'],
    };
    return nuevo as T;
  }

  // --- Empresas ---
  if (pathname === '/api/empresas') {
    if (method === 'POST') {
      const nueva: Empresa = {
        id: mockEmpresas.length + 1,
        nombre: String(body?.nombre ?? 'Cooperativa Demo'),
        telefono: (body?.telefono as string) ?? null,
        correo: (body?.correo as string) ?? null,
        tarifaEquipajeExtra: Number(body?.tarifaEquipajeExtra ?? 150),
        logoUrl: (body?.logoUrl as string) ?? null,
        activo: true,
      };
      return nueva as T;
    }
    return mockEmpresas as T;
  }

  if (pathname === '/api/empresas/mi-empresa') {
    return mockEmpresas[0] as T;
  }

  if (pathname === '/api/empresas/resumen-plataforma') {
    return mockResumenPlataforma as T;
  }

  const empresaDetalleId = matchId(pathname, '/api/empresas/');
  if (empresaDetalleId != null && pathname.endsWith('/detalle-plataforma')) {
    return mockDetalleCooperativa(empresaDetalleId) as T;
  }

  if (empresaDetalleId != null && pathname.endsWith('/desactivar') && method === 'PATCH') {
    return { message: 'Cooperativa desactivada (demo)', id: String(empresaDetalleId) } as T;
  }

  if (empresaDetalleId != null && method === 'PUT') {
    const emp = mockEmpresas.find((e) => e.id === empresaDetalleId) ?? mockEmpresas[0];
    return { ...emp, ...body } as T;
  }

  if (empresaDetalleId != null) {
    const emp = mockEmpresas.find((e) => e.id === empresaDetalleId);
    if (emp) return emp as T;
  }

  // --- Buses ---
  if (pathname === '/api/buses/mi-empresa') {
    return filterBuses(1, searchParams.get('sede')) as T;
  }

  if (pathname === '/api/buses' && method === 'GET') {
    const empresaId = searchParams.get('empresaId');
    return filterBuses(empresaId ? Number(empresaId) : null, searchParams.get('sede')) as T;
  }

  if (pathname === '/api/buses' && method === 'POST') {
    const bus: Bus = {
      id: mockBuses.length + 1,
      empresaId: Number(body?.empresaId ?? 1),
      numeroInterno: String(body?.numeroInterno ?? 'D-01'),
      placa: String(body?.placa ?? 'NI-9999'),
      capacidad: Number(body?.capacidad ?? 50),
      filas: 13,
      activo: true,
      fotoUrl: (body?.fotoUrl as string) ?? null,
      sede: String(body?.sede ?? 'Bluefields'),
    };
    return bus as T;
  }

  const busId = matchId(pathname, '/api/buses/');
  if (busId != null && pathname.includes('/asientos/') && method === 'PUT') {
    const asientoId = matchId(pathname, '/asientos/');
    const bus = mockBuses.find((b) => b.id === busId);
    const asiento = bus?.asientos?.find((a) => a.id === asientoId);
    return { ...asiento, ...body, id: asientoId ?? 1 } as T;
  }

  if (busId != null && method === 'PUT') {
    const bus = mockBuses.find((b) => b.id === busId) ?? mockBuses[0];
    return { ...bus, ...body } as T;
  }

  if (busId != null) {
    const bus = mockBuses.find((b) => b.id === busId);
    if (bus) return bus as T;
  }

  // --- Viajes ---
  if (pathname === '/api/viajes/mi-empresa') {
    const fecha = searchParams.get('fecha') ?? fechaHoyIso();
    return filterViajes(1, fecha, searchParams.get('origen')) as T;
  }

  if (pathname === '/api/viajes' && method === 'GET') {
    const fecha = searchParams.get('fecha') ?? fechaHoyIso();
    const empresaId = searchParams.get('empresaId');
    return filterViajes(empresaId ? Number(empresaId) : null, fecha, searchParams.get('origen')) as T;
  }

  if (pathname === '/api/viajes' && method === 'POST') {
    const fecha = String(body?.fecha ?? fechaHoyIso());
    const viaje: ViajeOperador = {
      id: 900 + mockViajesOperador(fecha).length,
      empresaId: Number(body?.empresaId ?? 1),
      empresaNombre: mockEmpresas.find((e) => e.id === Number(body?.empresaId))?.nombre ?? 'Wendelyn Transporte',
      busId: Number(body?.busId ?? 1),
      busNumeroInterno: mockBuses.find((b) => b.id === Number(body?.busId))?.numeroInterno ?? 'W-01',
      origen: String(body?.origen ?? 'Bluefields'),
      destino: String(body?.destino ?? 'Managua'),
      fecha,
      horaSalida: String(body?.horaSalida ?? '08:00:00'),
      tarifa: Number(body?.tarifa ?? 350),
      tarifaEquipajeExtra: Number(body?.tarifaEquipajeExtra ?? 150),
      estado: 'PROGRAMADO',
      asientosDisponibles: 50,
    };
    return viaje as T;
  }

  const viajeId = matchId(pathname, '/api/viajes/');
  if (viajeId != null && pathname.endsWith('/detalle-operador')) {
    const detalle = mockDetalleViajeOperador(viajeId);
    if (!detalle) throw new Error(`Viaje ${viajeId} no encontrado (demo)`);
    return detalle as T;
  }

  if (viajeId != null && pathname.endsWith('/cancelar') && method === 'PATCH') {
    const v = mockViajesOperador(fechaHoyIso()).find((x) => x.id === viajeId);
    return { ...v, estado: 'CANCELADO' } as T;
  }

  if (viajeId != null && method === 'PUT') {
    const v = mockViajesOperador(fechaHoyIso()).find((x) => x.id === viajeId);
    return { ...v, ...body } as T;
  }

  // --- Consulta pública ---
  if (pathname === '/api/publico/viajes' && method === 'GET') {
    const fecha = searchParams.get('fecha') ?? fechaHoyIso();
    const origen = searchParams.get('origen');
    const destino = searchParams.get('destino');
    let list = mockViajesDisponibles(fecha);
    if (origen && destino) {
      list = list.filter(
        (v) =>
          mockViajesOperador(fecha).some(
            (o) => o.id === v.viajeId && o.origen === origen && o.destino === destino
          )
      );
    }
    return list as T;
  }

  const viajePublicoId = matchId(pathname, '/api/publico/viajes/');
  if (viajePublicoId != null) {
    const detalle = mockDetalleViajePublico(viajePublicoId);
    if (!detalle) throw new Error(`Viaje ${viajePublicoId} no encontrado (demo)`);
    return detalle as T;
  }

  if (pathname.startsWith('/api/externo/tarifa-referencia-usd')) {
    const monto = Number(searchParams.get('monto') ?? 350);
    return { equivalenteUsd: Math.round((monto / 36.5) * 100) / 100 } as T;
  }

  // --- Ventas / pasajeros ---
  if (pathname === '/api/ventas' && method === 'POST') {
    const req = body as unknown as VentaRequest;
    return mockVentaResponse(req.viajeId, req.viajeAsientoIds) as T;
  }

  if (pathname === '/api/pasajeros/manifiesto') {
    const fecha = searchParams.get('fecha') ?? fechaHoyIso();
    let list = mockManifiesto(fecha);
    const viajeId = searchParams.get('viajeId');
    const busId = searchParams.get('busId');
    if (viajeId) list = list.filter((p) => p.viajeId === Number(viajeId));
    if (busId) {
      const bus = mockBuses.find((b) => b.id === Number(busId));
      if (bus) list = list.filter((p) => p.busNumeroInterno === bus.numeroInterno);
    }
    return list as T;
  }

  if (pathname === '/api/reservas-excepcionales' && method === 'POST') {
    return { id: 5001, numeroAsiento: 25 } as T;
  }

  // --- Reportes ---
  if (pathname === '/api/reportes/ocupacion') {
    const fecha = searchParams.get('fecha') ?? fechaHoyIso();
    let list = mockOcupacion(fecha);
    const empresaId = searchParams.get('empresaId');
    if (empresaId) {
      const ids = mockViajesOperador(fecha)
        .filter((v) => v.empresaId === Number(empresaId))
        .map((v) => v.id);
      list = list.filter((o) => ids.includes(o.viajeId));
    }
    return list as T;
  }

  if (pathname === '/api/reportes/ingresos') {
    const desde = searchParams.get('desde') ?? fechaHoyIso();
    const hasta = searchParams.get('hasta') ?? fechaHoyIso();
    return mockIngresos(desde, hasta) as T;
  }

  // --- Paradas ---
  if (pathname === '/api/paradas' && method === 'GET') {
    return mockParadas as T;
  }

  if (pathname === '/api/paradas' && method === 'POST') {
    const parada: ParadaRuta = {
      id: mockParadas.length + 1,
      nombre: String(body?.nombre ?? 'Parada demo'),
      orden: mockParadas.length + 1,
      minutosDesdeSalida: Number(body?.minutosDesdeSalida ?? 0),
      horaEstimada: '08:00:00',
      latitud: Number(body?.latitud ?? 12.0),
      longitud: Number(body?.longitud ?? -86.0),
    };
    return parada as T;
  }

  const paradaId = matchId(pathname, '/api/paradas/');
  if (paradaId != null && method === 'PUT') {
    const p = mockParadas.find((x) => x.id === paradaId) ?? mockParadas[0];
    return { ...p, ...body } as T;
  }

  if (paradaId != null && method === 'DELETE') {
    return undefined as T;
  }

  throw new Error(`[Modo demo] Ruta no simulada: ${method} ${pathname}`);
}

File: mobile/lib/mocks/mock_api.dart
import '../core/api/api_client.dart';
import '../core/utils/formato.dart';
import 'mock_data.dart';
import 'mock_demo_profile.dart';

Future<void> _delay() => Future<void>.delayed(const Duration(milliseconds: mockDelayMs));

int? _matchId(String pathname, String prefix) {
  if (!pathname.startsWith(prefix)) return null;
  final rest = pathname.substring(prefix.length);
  final idStr = rest.split('/').first;
  return int.tryParse(idStr);
}

List<Map<String, dynamic>> _filterViajes({
  required String fecha,
  int? empresaId,
  String? origen,
}) {
  var list = mockViajesOperadorJson(fecha);
  if (empresaId != null) {
    list = list.where((v) => v['empresaId'] == empresaId).toList();
  }
  if (origen != null && origen.isNotEmpty) {
    list = list.where((v) => v['origen'] == origen).toList();
  }
  return list;
}

List<Map<String, dynamic>> _filterBuses({int? empresaId, String? sede}) {
  var list = mockBusesJson.cast<Map<String, dynamic>>().toList();
  if (empresaId != null) {
    list = list.where((b) => b['empresaId'] == empresaId).toList();
  }
  if (sede != null && sede.isNotEmpty) {
    list = list.where((b) => b['sede'] == sede).toList();
  }
  return list;
}

Future<dynamic> handleMockRequest(
  String method,
  String path, {
  Map<String, String>? query,
  Object? body,
}) async {
  await _delay();

  final q = query ?? {};
  final bodyMap = switch (body) {
    Map<String, dynamic> m => m,
    Map m => Map<String, dynamic>.from(m),
    _ => null,
  };
  final pathname = path.split('?').first;

  // Perfil / usuarios
  if (pathname == '/api/usuarios/me') return mockPerfilForProfile(MockSession.profile);

  if (pathname == '/api/usuarios' && method == 'GET') {
    final empresaId = q['empresaId'];
    if (empresaId != null) {
      return mockOperadoresJson.where((o) => o['empresaId'] == int.parse(empresaId)).toList();
    }
    return mockOperadoresJson;
  }

  if (pathname == '/api/usuarios' && method == 'POST') {
    return {
      'id': 99,
      'empresaId': bodyMap?['empresaId'] ?? 1,
      'empresaNombre': 'Wendelyn Transporte',
      'nombreUsuario': bodyMap?['nombreUsuario'] ?? 'nuevo.operador',
      'emailLogin': bodyMap?['email'] ?? 'nuevo@demo.com',
      'nombreCompleto': bodyMap?['nombreCompleto'] ?? 'Operador Demo',
      'sede': bodyMap?['sede'] ?? 'Bluefields',
      'activo': true,
      'roles': bodyMap?['roles'] ?? ['CAJERO'],
    };
  }

  final usuarioId = _matchId(pathname, '/api/usuarios/');
  if (usuarioId != null && method == 'PATCH') {
    final user = mockOperadoresJson.cast<Map<String, dynamic>>().firstWhere(
          (o) => o['id'] == usuarioId,
          orElse: () => mockPerfilForProfile(MockSession.profile),
        );
    return {...user, ...?bodyMap};
  }

  // Empresas
  if (pathname == '/api/empresas' && method == 'GET') return mockEmpresasJson;

  if (pathname == '/api/empresas' && method == 'POST') {
    return {
      'id': mockEmpresasJson.length + 1,
      'nombre': bodyMap?['nombre'] ?? 'Cooperativa Demo',
      'telefono': bodyMap?['telefono'],
      'correo': bodyMap?['correo'],
      'tarifaEquipajeExtra': bodyMap?['tarifaEquipajeExtra'] ?? 150.0,
      'logoUrl': bodyMap?['logoUrl'],
      'activo': true,
    };
  }

  if (pathname == '/api/empresas/mi-empresa') return mockEmpresasJson.first;

  if (pathname == '/api/empresas/resumen-plataforma') return mockResumenPlataformaJson;

  final empresaId = _matchId(pathname, '/api/empresas/');
  if (empresaId != null && pathname.endsWith('/detalle-plataforma')) {
    return mockDetalleCooperativaJson(empresaId);
  }

  if (empresaId != null && pathname.endsWith('/desactivar') && method == 'PATCH') {
    return {'message': 'Cooperativa desactivada (demo)', 'id': '$empresaId'};
  }

  if (empresaId != null && method == 'PUT') {
    final emp = mockEmpresasJson.cast<Map<String, dynamic>>().firstWhere(
          (e) => e['id'] == empresaId,
          orElse: () => mockEmpresasJson.first,
        );
    return {...emp, ...?bodyMap};
  }

  if (empresaId != null) {
    for (final e in mockEmpresasJson) {
      if (e['id'] == empresaId) return e;
    }
  }

  // Buses
  if (pathname == '/api/buses/mi-empresa') {
    return _filterBuses(empresaId: 1, sede: q['sede']);
  }

  if (pathname == '/api/buses' && method == 'GET') {
    final empId = q['empresaId'] != null ? int.parse(q['empresaId']!) : null;
    return _filterBuses(empresaId: empId, sede: q['sede']);
  }

  if (pathname == '/api/buses' && method == 'POST') {
    return {
      'id': mockBusesJson.length + 1,
      'empresaId': bodyMap?['empresaId'] ?? 1,
      'numeroInterno': bodyMap?['numeroInterno'] ?? 'D-01',
      'placa': bodyMap?['placa'] ?? 'NI-9999',
      'capacidad': bodyMap?['capacidad'] ?? 50,
      'filas': 13,
      'activo': true,
      'fotoUrl': bodyMap?['fotoUrl'],
      'sede': bodyMap?['sede'] ?? 'Bluefields',
      'asientos': _buildAsientosBusJson(mockBusesJson.length + 1),
    };
  }

  final busId = _matchId(pathname, '/api/buses/');
  if (busId != null && pathname.contains('/asientos/') && method == 'PUT') {
    final asientoId = _matchId(pathname, '/asientos/');
    return {
      'id': asientoId,
      'numero': bodyMap?['numero'] ?? 1,
      'fila': bodyMap?['fila'] ?? 1,
      'posicion': bodyMap?['posicion'] ?? 'VENTANA',
    };
  }

  if (busId != null && method == 'PUT') {
    final bus = mockBusesJson.cast<Map<String, dynamic>>().firstWhere(
          (b) => b['id'] == busId,
          orElse: () => mockBusesJson.first,
        );
    return {...bus, ...?bodyMap};
  }

  if (busId != null) {
    for (final b in mockBusesJson) {
      if (b['id'] == busId) return b;
    }
  }

  // Viajes
  if (pathname == '/api/viajes/mi-empresa') {
    return _filterViajes(
      fecha: q['fecha'] ?? fechaHoyIso(),
      empresaId: 1,
      origen: q['origen'],
    );
  }

  if (pathname == '/api/viajes' && method == 'GET') {
    return _filterViajes(
      fecha: q['fecha'] ?? fechaHoyIso(),
      empresaId: q['empresaId'] != null ? int.parse(q['empresaId']!) : null,
      origen: q['origen'],
    );
  }

  if (pathname == '/api/viajes' && method == 'POST') {
    final fecha = bodyMap?['fecha']?.toString() ?? fechaHoyIso();
    final empId = bodyMap?['empresaId'] as int? ?? 1;
    final busIdVal = bodyMap?['busId'] as int? ?? 1;
    Map<String, dynamic>? bus;
    for (final b in mockBusesJson) {
      if (b['id'] == busIdVal) {
        bus = b;
        break;
      }
    }
    Map<String, dynamic>? emp;
    for (final e in mockEmpresasJson) {
      if (e['id'] == empId) {
        emp = e;
        break;
      }
    }
    return {
      'id': 900 + mockViajesOperadorJson(fecha).length,
      'empresaId': empId,
      'empresaNombre': emp?['nombre'] ?? 'Wendelyn Transporte',
      'busId': busIdVal,
      'busNumeroInterno': bus?['numeroInterno'] ?? 'W-01',
      'origen': bodyMap?['origen'] ?? 'Bluefields',
      'destino': bodyMap?['destino'] ?? 'Managua',
      'fecha': fecha,
      'horaSalida': bodyMap?['horaSalida'] ?? '08:00:00',
      'tarifa': bodyMap?['tarifa'] ?? 350.0,
      'tarifaEquipajeExtra': bodyMap?['tarifaEquipajeExtra'] ?? 150.0,
      'estado': 'PROGRAMADO',
      'asientosDisponibles': 50,
    };
  }

  final viajeId = _matchId(pathname, '/api/viajes/');
  if (viajeId != null && pathname.endsWith('/detalle-operador')) {
    final detalle = mockDetalleViajeJson(viajeId);
    if (detalle == null) throw ApiException('Viaje $viajeId no encontrado (demo)');
    return detalle;
  }

  if (viajeId != null && pathname.endsWith('/cancelar') && method == 'PATCH') {
    final v = mockViajesOperadorJson(fechaHoyIso()).firstWhere(
      (x) => x['id'] == viajeId,
      orElse: () => mockViajesOperadorJson(fechaHoyIso()).first,
    );
    return {...v, 'estado': 'CANCELADO'};
  }

  if (viajeId != null && method == 'PUT') {
    final v = mockViajesOperadorJson(fechaHoyIso()).firstWhere(
      (x) => x['id'] == viajeId,
      orElse: () => mockViajesOperadorJson(fechaHoyIso()).first,
    );
    return {...v, ...?bodyMap};
  }

  // Consulta pública
  if (pathname == '/api/publico/viajes' && method == 'GET') {
    final fecha = q['fecha'] ?? fechaHoyIso();
    return mockViajesDisponiblesJson(
      fecha,
      origen: q['origen'],
      destino: q['destino'],
    );
  }

  final viajePublicoId = _matchId(pathname, '/api/publico/viajes/');
  if (viajePublicoId != null) {
    final detalle = mockDetalleViajeJson(viajePublicoId);
    if (detalle == null) throw ApiException('Viaje $viajePublicoId no encontrado (demo)');
    return detalle;
  }

  if (pathname.startsWith('/api/externo/tarifa-referencia-usd')) {
    final monto = double.tryParse(q['monto'] ?? '350') ?? 350;
    return {'equivalenteUsd': double.parse((monto / 36.5).toStringAsFixed(2))};
  }

  // Ventas / pasajeros
  if (pathname == '/api/ventas' && method == 'POST') {
    return mockVentaResponseJson(bodyMap ?? {});
  }

  if (pathname == '/api/pasajeros/manifiesto') {
    final fecha = q['fecha'] ?? fechaHoyIso();
    var list = mockManifiestoJson(fecha);
    if (q['viajeId'] != null) {
      final id = int.parse(q['viajeId']!);
      list = list.where((p) => p['viajeId'] == id).toList();
    }
    if (q['busId'] != null) {
      final busIdVal = int.parse(q['busId']!);
      Map<String, dynamic>? bus;
      for (final b in mockBusesJson) {
        if (b['id'] == busIdVal) {
          bus = b;
          break;
        }
      }
      if (bus != null) {
        list = list.where((p) => p['busNumeroInterno'] == bus!['numeroInterno']).toList();
      }
    }
    return list;
  }

  if (pathname == '/api/reservas-excepcionales' && method == 'POST') {
    return {'id': 5001, 'numeroAsiento': 25};
  }

  // Reportes
  if (pathname == '/api/reportes/ocupacion') {
    final fecha = q['fecha'] ?? fechaHoyIso();
    var list = mockOcupacionJson(fecha);
    if (q['empresaId'] != null) {
      final empId = int.parse(q['empresaId']!);
      final ids = mockViajesOperadorJson(fecha)
          .where((v) => v['empresaId'] == empId)
          .map((v) => v['id'])
          .toSet();
      list = list.where((o) => ids.contains(o['viajeId'])).toList();
    }
    return list;
  }

  if (pathname == '/api/reportes/ingresos') {
    final desde = q['desde'] ?? fechaHoyIso();
    final hasta = q['hasta'] ?? fechaHoyIso();
    return mockIngresosJson(desde, hasta);
  }

  // Paradas
  if (pathname == '/api/paradas' && method == 'GET') {
    final origen = q['origen'];
    final destino = q['destino'];
    if (origen != null && destino != null) {
      return mockParadasForRoute(origen, destino);
    }
    return mockParadasBfsMgaJson;
  }

  if (pathname == '/api/paradas' && method == 'POST') {
    return {
      'id': mockParadasBfsMgaJson.length + 1,
      'nombre': bodyMap?['nombre'] ?? 'Parada demo',
      'orden': mockParadasBfsMgaJson.length + 1,
      'minutosDesdeSalida': bodyMap?['minutosDesdeSalida'] ?? 0,
      'horaEstimada': '08:00:00',
      'latitud': bodyMap?['latitud'] ?? 12.0,
      'longitud': bodyMap?['longitud'] ?? -86.0,
    };
  }

  final paradaId = _matchId(pathname, '/api/paradas/');
  if (paradaId != null && method == 'PUT') {
    final p = mockParadasBfsMgaJson.cast<Map<String, dynamic>>().firstWhere(
          (x) => x['id'] == paradaId,
          orElse: () => mockParadasBfsMgaJson.first,
        );
    return {...p, ...?bodyMap};
  }

  if (paradaId != null && method == 'DELETE') return null;

  throw ApiException('[Modo demo] Ruta no simulada: $method $pathname');
}

List<Map<String, dynamic>> _buildAsientosBusJson(int busId) {
  final asientos = <Map<String, dynamic>>[];
  var numero = 1;
  for (var fila = 1; fila <= 12; fila++) {
    for (final pos in ['VENTANA', 'PASILLO']) {
      if (numero > 45) break;
      asientos.add({
        'id': busId * 100 + numero,
        'numero': numero,
        'fila': fila,
        'posicion': pos,
      });
      numero++;
    }
  }
  for (var i = 1; i <= 5; i++) {
    asientos.add({
      'id': busId * 100 + numero,
      'numero': numero,
      'fila': 13,
      'posicion': 'TRASERA_$i',
    });
    numero++;
  }
  return asientos;
}
